### DISCUSSION TITLE
Can I use Nickel in a Nix expression?

### DISCUSSION THREAD

**Discussion by akavel** [tweag/nickel#1769] [Q&A] [answered]
It's not really clear to me now: is there a way, when writing a normal Nix flake, for me to transparently use Nickel to write as much as possible of it?

I assume the answer is unfortunately a "no", but I can't be sure until I ask. I assume I can currently only use it in a two-step process, manually pre-generating some JSON with Nickel, committing it into the flake repo, and then having it imported in the Nix flake.

**Comment by yannham** [comment, marked as answer]
I think your best bet might be the [organist project](https://github.com/nickel-lang/organist), which we also develop at Tweag. It's mostly centered around writing development shells, so depending on what you try to do, it might not be the right solution, but it's probably the most advanced example of Nickel-Nix integration today. Organist have mechanisms to pass derivations between Nix and Nickel (of course, there are many restrictions, but using simple packages without overriding shouldn't be an issue), for example.

Even if organist, as it is, doesn't fit your use-case, you might be able to steal ideas and code from its internal library. Organist has a Nix library with functions such as `importFromNcl` which handle a bunch of task (running Nickel, generating a JSON, importing it back, parsing the result according the organist conventions to transform some values with specific markers back to actual Nix derivations, etc.), and you might be able to tweak it or even just use it directly to interface with Nickel using a single Nix function as an entry point.

I haven't followed as closely organist development lately, so if you have any technical question specifically related to organist, you might have more luck asking directly there.

**Reply by akavel** [reply]
Thanks! Esp. for the mention of `importFromNcl`, sounds like something I should give a look. I am interested in using Nickel in configs for home-manager and maybe NixOS, NixOps and nix-darwin as well (possibly via https://github.com/srid/nixos-flake), so I don't suspect Organist would cover those usecases.

**Reply by yannham** [reply]
I see. I've been wanting to take a (experimental) stab at writing NixOS configs as well, but haven't had the time yet. I think writing a Nickel file that sets some simple module options shouldn't be too hard (after all, you can probably do that in JSON). Interacting in non trivial ways with NixOS modules might be a bit harder.

One thing that would be very cool, though, would be to import all the knowledge of NixOS module types into Nickel. That is, generate Nickel contracts and metadata from module descriptions to have them in the LSP and have the nice contract error reporting. IIRC you can spit out NixOS module options descriptions as JSON, so I think it's feasible to process this file and generate Nickel contracts automatically - in the same vein as what we're doing for [json-schema-to-nickel](https://github.com/nickel-lang/json-schema-to-nickel/) or for auto-generating Terraform contracts in [nickel-terraform](https://github.com/tweag/tf-ncl). But this does require some work to reach that stage, though...
